src/common/getType.ts

```diff
@@ -579,60 +717,17 @@ export function getCollectionTypes(expr: Expr, program: Program): ValueType[] {
   throw new PolygolfError("Type error. Node is not a collection.", expr.source);
 }
 
-/*
-function floorDiv(a: bigint, b: bigint): bigint {
-  const res = a / b;
-  return a < 0 !== b < 0 ? res - 1n : res;
+function getIntegerTypeMod(a: IntegerType, b: IntegerType): IntegerType {
+  const values: IntegerBound[] = [];
+  if (lt(b.low, 0n))
+    values.push(sub(1n, min(b.low, min(abs(a.low), abs(a.high)))));
+  if (lt(0n, b.high))
+    values.push(sub(min(b.low, min(abs(a.low), abs(a.high))), 1n));
+  values.push(0n);
+  return integerTypeIncludingAll(...values);
 }
-*/
 
-/** Combines types `left` and `right` using the *convex* operator `op` */
-function getIntegerTypeUsing(
-  left: IntegerType,
-  right: IntegerType,
-  op: (a: bigint, b: bigint) => bigint
-): IntegerType {
-  if (
-    left.low === undefined ||
-    left.high === undefined ||
-    right.low === undefined ||
-    right.high === undefined
-  )
-    return integerType();
-  return integerTypeIncludingAll([
-    op(left.low, right.low),
-    op(left.low, right.high),
-    op(left.high, right.low),
-    op(left.high, right.high),
-  ]);
-}
-
-function getIntegerTypeMod(left: IntegerType, right: IntegerType): IntegerType {
-  if (right.low === undefined || right.high === undefined) {
-    if (left.low === undefined || left.high === undefined) {
-      return integerType();
-    }
-    const m = left.high > -left.low ? left.high : -left.low;
-    return integerType(-m, m);
-  }
-  if (
-    left.low === undefined ||
-    left.high === undefined ||
-    left.low <= 0n ||
-    left.high >= right.high
-  ) {
-    const values = [0n];
-    if (right.low < 0n) values.push(right.low + 1n);
-    if (right.high > 0n) values.push(right.high - 1n);
-    return integerTypeIncludingAll(values);
-  }
-  return left;
-}
-
-function getIntegerTypeRem(left: IntegerType, right: IntegerType): IntegerType {
-  if (right.low === undefined || right.high === undefined) {
-    return integerType();
-  }
-  const m = right.high > -right.low ? right.high : -right.low;
-  return integerType(-m, m);
+function getIntegerTypeRem(a: IntegerType, b: IntegerType): IntegerType {
+  const m = max(abs(b.low), abs(b.high));
+  return integerType(neg(m), m);
```

Comment:
Might be useful to add in an extra case: if `a` cannot be negative, then the remainder cannot be negative.